keps/sig-auth/20190313-dry-run-authz.md

```diff
@@ -0,0 +1,77 @@
+---
+title: Dry-run authz
+authors:
+  - "@apelisse"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - "@liggitt"
+  - "@lavalamp"
+editor: TBD
+creation-date: 2019-03-13
+last-updated: 2019-03-13
+status: provisional
+see-also:
+  - "/keps/sig-api-machinery/0015-dry-run.md"
+replaces:
+superseded-by:
+---
+
+# Dry-run authz
+
+## Table of Contents
+
+## Summary
+
+Kubernetes server-side dry-run is a feature that lets people validate their
+changes without having to persist them. It allows them to verify that the
+validation (including built-in admission chain and validating webhooks) for
+their object would be passing, and to know what changes would be made by
+defaulter and mutating webhooks. It's also used as a foundation to build
+`kubectl diff` which lets people visualize their changes in a more natural
+way. It's a powerful tool for CI/CD. Unfortunately, dry-run doesn't have a
+specific authorization model, which means that users need write-access to
+perform these non-persisting requests.
```

Comment:
authorizing dry-run differently doesn't seem correct to me. it makes the results of dry run even more detached from reality.

do we have evidence of people using this approach for CI/CD? I'm much more familiar with validating manifests against a test cluster where the CI system actually has write permission, and the deployment process mirrors that used in production

Reply:
Alternative proposal:
Add a regular user with write access. Add an authz plugin / admission plugin that rejects every request this user makes that doesn't specify "dry-run".

I don't know how to mitigate the risk already mentioned below with this system, either. But it does less violence to the system, maybe.

Reply:
I'm also somewhat skeptical that a CI system could generically use dry run to successfully validate entire collections of objects involved in a deployment. Something as simple as "create namespace foo, then create objects inside that namespace" would fail on dry-run, since the object creations would be rejected because the namespace didn't actually get created.